ziro-export-risport.ps1

```diff
@@ -0,0 +1,227 @@
+      $totalDevicesFound = $response.Envelope.Body.selectCmDeviceResponse.selectCmDeviceReturn.SelectCmDeviceResult.TotalDevicesFound
+
+      if ($totalDevicesFound -gt 0) {
+         $devicesPerNode = $response.Envelope.Body.selectCmDeviceResponse.selectCmDeviceReturn.SelectCmDeviceResult.CmNodes.item.CmDevices 
+         | Where-Object { $_.HasChildNodes }
+
+         foreach ($nodeDevices in $devicesPerNode) {
+            $nodeDevices.item | Select-Object -Property Name, Status, StatusReason, TimeStamp 
+            | ForEach-Object { $registrationStatuses.Add($_.Name, $_) }
+         }
+      }
+
+      $windowStartIndex = $windowStartIndex + $maxReturnedDevices
+      Start-Sleep -Seconds 1  # a pinch of throttling to help with rate limits
+   }
+
+   return $registrationStatuses
+}
+
+
+$serverUrl = Read-Host 'CUCM server URL (https://mycucm.company.com)'
+$credential = Get-Credential -Message 'Enter username and password'
+$cucmConnector = [CucmConnector]::new($serverUrl, $credential) 
+
+try {
+   Write-Host "Fetching devices from CUCM server..."
+   $devices = Get-Devices -CucmConnector $cucmConnector
+   Write-Host "Found [$($devices.Length)] devices"
+   Write-Host "Getting device registration statuses..."
+   $registrationStatuses = Get-DeviceRegistrationStatuses -DeviceNames $devices.name -CucmConnector $cucmConnector
+   $exportResults = New-Object System.Collections.Generic.List[Hashtable]
+
+   foreach ($device in $devices) {
+      if ($registrationStatuses.ContainsKey($device.name)) {
+         $matchingStatus = $registrationStatuses[$device.name]
+         $exportResults.Add(@{
+               Name         = $device.name
+               Status       = $matchingStatus.Status
+               StatusReason = $matchingStatus.StatusReason
+               TimeStamp    = $matchingStatus.TimeStamp
+               Model        = $device.model
+            })
+      }
+      else {
+         $exportResults.Add(@{
+               Name         = $device.name
+               Status       = "Unknown"
+               StatusReason = ""
+               TimeStamp    = ""
+               Model        = $device.model
+            })
+      }
+   }
+
+   # Export to JSON file
+   $serverHost = ([System.Uri]$serverUrl).Host
+   $date = (Get-Date -Format 'yyyy-MM-dd_HH-mm-ss').ToString()
+   $outputFileName = "${date}_${serverHost}.json"
+   $outputFile = $exportResults | ConvertTo-Json -depth 100 -AsArray | New-Item -Path . -Name $outputFileName -ItemType File
+   Write-Host "RISport information export was successful: $outputFile"
+   exit 0
+}
+catch {
+   $responseCode = $_.Exception.Response.StatusCode.value__
+
+   if ($responseCode -eq 401) {
+      Write-Error "Invalid credentials (401 Unauthorized): $_"
+   }
+   elseif ($responseCode -eq 403) {
+      Write-Error "Insufficient permissions (403 Forbidden): $_"
+   }
+   else {
+      Write-Error "Ran into errors when exporting RISport information: $_"
+   }
+
+   exit 1
```

Comment:
Let's add a failure message before terminating to better align with the story requirements.

Reply:
👍